File: domox-domain/src/main/java/domox/SimpleModule.java
package domox;

import domox.dom.rqm.Author;
import lombok.Data;
import org.apache.isis.testing.fixtures.applib.fixturescripts.FixtureScript;
import org.apache.isis.testing.fixtures.applib.modules.ModuleWithFixtures;
import org.springframework.boot.autoconfigure.domain.EntityScan;
import org.springframework.context.annotation.ComponentScan;
import org.springframework.context.annotation.Configuration;
import org.springframework.data.jpa.repository.config.EnableJpaRepositories;
import org.springframework.validation.annotation.Validated;

@Configuration
/*@Import({
        IsisModuleCoreRuntimeServices.class,
        IsisModuleTestingFixturesApplib.class,
})
@PropertySources({
        @PropertySource(IsisPresets.NoTranslations),
        @PropertySource(IsisPresets.SilenceWicket),
})*/
@ComponentScan
@EnableJpaRepositories
@EntityScan(basePackageClasses = {SimpleModule.class})
public class SimpleModule implements ModuleWithFixtures {

    @Override
    public FixtureScript getTeardownFixture() {
        return new FixtureScript() {
            @Override
            protected void execute(ExecutionContext executionContext) {
                repositoryService.removeAll(Author.class);
            }
        };
    }

    public static class PropertyDomainEvent<S, T>
            extends org.apache.isis.applib.events.domain.PropertyDomainEvent<S, T> {
    }

    public static class CollectionDomainEvent<S, T>
            extends org.apache.isis.applib.events.domain.CollectionDomainEvent<S, T> {
    }

    public static class ActionDomainEvent<S>
            extends org.apache.isis.applib.events.domain.ActionDomainEvent<S> {
    }

    @Data
    @Validated
    public static class Configuration {
        private final Types types = new Types();

        @Data
        public static class Types {
            private final Name name = new Name();

            @Data
            public static class Name {
                private final Validation validation = new Validation();

                @Data
                public static class Validation {
                    private char[] prohibitedCharacters = "!&%$".toCharArray();
                    private String message = "Character '{character}' is not allowed";
                }
            }
        }
    }
}

File: module-ksimple/src/main/kotlin/org.apache.isis.domox/modules/ksimple/SimpleModule.kt
package org.apache.isis.domox.modules.ksimple

import org.apache.isis.domox.modules.ksimple.dom.so.SimpleObject
import org.apache.isis.testing.fixtures.applib.modules.ModuleWithFixtures
import org.apache.isis.testing.fixtures.applib.teardown.TeardownFixtureAbstract
import org.springframework.boot.context.properties.ConfigurationProperties
import org.springframework.boot.context.properties.EnableConfigurationProperties
import org.springframework.context.annotation.ComponentScan
import org.springframework.context.annotation.Configuration
import org.springframework.context.annotation.Import
import org.springframework.validation.annotation.Validated

@Configuration
@Import()
@ComponentScan
@EnableConfigurationProperties(Configuration::class)
class SimpleModule : ModuleWithFixtures {
    // @get:Override
    val teardownFixture = object : TeardownFixtureAbstract() {
        @Override
        override fun execute(executionContext: ExecutionContext?) {
            deleteFrom(SimpleObject::class.java)
        }
    }

    open class PropertyDomainEvent<S, T> : org.apache.isis.applib.events.domain.PropertyDomainEvent<S, T>()
    open class CollectionDomainEvent<S, T> : org.apache.isis.applib.events.domain.CollectionDomainEvent<S, T>()
    open class ActionDomainEvent<S> : org.apache.isis.applib.events.domain.ActionDomainEvent<S>()

    @ConfigurationProperties("app.ksimple-module")
    @Validated
    class Configuration {
        private val types = Types()
        fun getTypes(): Types {
            return types
        }

        class Types {
            private val name = Name()
            fun getName(): Name {
                return name
            }

            class Name {
                private val validation = Validation()
                fun getValidation(): Validation {
                    return validation
                }

                class Validation {
                    private val prohibitedCharacters = "!&%$".toCharArray()
                    fun getProhibitedCharacters(): CharArray {
                        return prohibitedCharacters
                    }

                    private val message = "Character '{character}' is not allowed"
                    fun getMessage(): String {
                        return message
                    }
                }
            }
        }
    }
}
